### PR TITLE
Add services for pre and post runh of sapinit

### PR DESCRIPTION
**Pull request by fmherschel**
* fix for SAPStartSrv to give correct rc for monitor probe
* adding pre and post service for sapinit
* man pages

### CODE REVIEW

**Comment by fmherschel** [comment]
@arbulu89 sapservices-move is a very short shell script. It is called via pre and post services sapping and sappong which are running before/after sapinit.
If we like to translate it from shell to python I would like to do it myself.

File: sapstartsrv-resource-agents.spec [hunk 1]
```diff
@@ -49,13 +47,9 @@ BuildRequires:  python3-pytest
...
 
 %description
-This project is to implement a resource agent for the instance specific SAP start framework. It controls the instance specific sapstartsrv process which provides the API to start, stop and check an SAP instance.
-
-SAPStartSrv does only start, stop and probe for the server process. By intention it does not monitor the service. SAPInstance is doing in-line recovery of failed sapstartsrv processes instead.
-
-SAPStartSrv is to be included into a resource group together with the vIP and the SAPInstance. It needs to be started before SAPInstance is starting and needs to be stopped after SAPInstance has been stopped.
-
-SAPStartSrv can be used since SAP NetWeaver 7.40 or SAP S/4HANA (ABAP Platform >= 1909).
+This is an resource agent for the instance specific SAP start framework.
```
Comment (arbulu89) [inline sapstartsrv-resource-agents.spec:50]: `/This is an resource/This is a resource`

File: sapstartsrv-resource-agents.spec [hunk 2]
```diff
@@ -71,9 +65,15 @@ gzip man/*
...
 mkdir -p %{buildroot}/usr/lib/ocf/resource.d/suse
 mkdir -p %{buildroot}%{_mandir}/man7
+mkdir -p %{buildroot}%{_mandir}/man8
+mkdir -p %{buildroot}/usr/sbin
+mkdir -p %{buildroot}/usr/lib/systemd/system
 
 install -m 0755 ra/%{raname}.in %{buildroot}/usr/lib/ocf/resource.d/suse/%{raname}
 install -m 0444 man/*.7.gz %{buildroot}%{_mandir}/man7
+install -m 0444 man/*.8.gz %{buildroot}%{_mandir}/man8
+install -m 0755 sbin/* %{buildroot}/usr/sbin
```
Comment (arbulu89) [inline sapstartsrv-resource-agents.spec:75]: Do we need to give `write` permissions to this file? It shouldn't be `644` maybe?

File: sapstartsrv-resource-agents.spec [hunk 3]
```diff
@@ -71,9 +65,15 @@ gzip man/*
 %install
 mkdir -p %{buildroot}/usr/lib/ocf/resource.d/suse
 mkdir -p %{buildroot}%{_mandir}/man7
+mkdir -p %{buildroot}%{_mandir}/man8
+mkdir -p %{buildroot}/usr/sbin
```
Comment (arbulu89) [inline sapstartsrv-resource-agents.spec:69]: I don't think we need to create the `sbin` and `systemd` folders. There are already macros and shortcuts for that like `%{_sbindidr}` or `%{_unitdir}`.
Actually, I would recommend to check how other systemd services are shipped. It doesn't look like this how it should be done.

**Review by arbulu89 — COMMENTED** [review]
@fmherschel Some thoughts:
- I think the fix for `monitor` method should come in a different PR. They are different implementations and should be treated as such. Otherwise, it will be difficult to find things in the future. Besides that, the test coverage is not 100% anymore, so some code path is not tested.
- I don't like we start mixing Python and bash again, and the bash code doesn't have any kind of testing.
- I don't understand why we create the ping/hide - pong/restore naming. Why not using `hide/restore` for everything.
- About the systemd service themselves, I cannot really judge, but they look like a workaround. But maybe there is not any other alternative. As we are working closely with SAP, couldn't we talk with them to see if we can change the `sapinit` service itself to do these operations?

**Comment by lpinne** [comment]
> @fmherschel Some thoughts:
> 
> * I think the fix for `monitor` method should come in a different PR. They are different implementations and should be treated as such. Otherwise, it will be difficult to find things in the future. Besides that, the test coverage is not 100% anymore, so some code path is not tested.

@arbulu89:
> * I don't like we start mixing Python and bash again, and the bash code doesn't have any kind of testing.
Good point.
The shell script was used only to check if the concept works. Fortunately it did and we can proceed. No need to stick with bash.

> * I don't understand why we create the ping/hide - pong/restore naming. Why not using `hide/restore` for everything.
The systemd  services are called sapping/sappong. So the related functions are named similar.

> * About the systemd service themselves, I cannot really judge, but they look like a workaround. But maybe there is not any other alternative. As we are working closely with SAP, couldn't we talk with them to see if we can change the `sapinit` service itself to do these operations?
It is a workaround for sapinit. As the SLE-HA product is just a workaround for failing hardware and unreliable software. Indeed we have created a ticket for discussing with SAP if they could improve the next generation of saphostagent. If we are lucky they will look into that in the next couple of years.

**Comment by lpinne** [comment]
Hallo Fabian,

was soll ich machen?


Viele Grüße,

Lars

fmherschel wrote:

File: sapstartsrv-resource-agents.spec (hunk 2 above)
Comment (fmherschel) [reply sapstartsrv-resource-agents.spec:75]: AAARRRRRGGGLLLLLL !! In the past I one time did package a file with 644 (script) and got a longish discussion
why I limit root, if root could however change the file ;-)

**Review by fmherschel — COMMENTED** [review]
> @arbulu89 sapservices-move is a very short shell script. It is called via pre and post services sapping and sappong which are running before/after sapinit.
> If we like to translate it from shell to python I would like to do it myself.

I will redraw the PR and come back with a python script.

**Pull request closed without merging**